packages/portalnetwork/src/wire/utp/PortalNetworkUtp/requestManager.ts

```diff
@@ -50,43 +61,52 @@ export class RequestManager {
+            await request.handleUtpPacket(packet)
+            return
         } else {
-            this.masterPacketQueue.push(packet)
+            this.packetHeap.push(packet)
         }
-        this.logger.extend('HANDLE_PACKET')(`Adding ${PacketType[packet.header.pType]} packet for request ${packet.header.connectionId} to packet queue (size: ${this.masterPacketQueue.length} packets)`)
+        this.logger.extend('HANDLE_PACKET')(`Adding ${PacketType[packet.header.pType]} [${packet.header.seqNr}] for Req:${packet.header.connectionId} to queue (size: ${this.packetHeap.size()} packets)`)
         if (this.currentPacket === undefined) {
-            this.currentPacket = this.masterPacketQueue.shift()
+            this.currentPacket = this.packetHeap.pop()
             await this.processCurrentPacket()
         }
     }
 
-    async processCurrentPacket() {
-        this.logger.extend('PROCESS_CURRENT_PACKET')(`Packet Queue Size: ${this.masterPacketQueue.length}`)
+    async processCurrentPacket(): Promise<void> {
+        this.logger.extend('PROCESS_CURRENT_PACKET')(`Packet Queue Size: ${this.packetHeap.size()}`)
         if (this.currentPacket === undefined) {
-            if (this.masterPacketQueue.length === 0) {
+            if (this.packetHeap.size() === 0) {
                 this.logger.extend('PROCESS_CURRENT_PACKET')(`No packets to process`)
                 return
             }
-            this.currentPacket = this.masterPacketQueue.shift()
+            this.currentPacket = this.packetHeap.pop()
             await this.processCurrentPacket()
             return
         }
-        this.logger.extend('PROCESS_CURRENT_PACKET')(`Processing ${PacketType[this.currentPacket.header.pType]} packet for request ${this.currentPacket.header.connectionId}`)
+        this.logger.extend('PROCESS_CURRENT_PACKET')(`Processing ${PacketType[this.currentPacket.header.pType]} [${this.currentPacket.header.seqNr}] for Req:${this.currentPacket.header.connectionId}`)
         const request = this.lookupRequest(this.currentPacket.header.connectionId)
         if (request === undefined) {
             this.logger.extend('PROCESS_CURRENT_PACKET')(`Request not found for current packet - connectionId: ${this.currentPacket.header.connectionId}`)
-            this.currentPacket = this.masterPacketQueue.shift()
+            this.currentPacket = this.packetHeap.pop()
             await this.processCurrentPacket()
             return
         }
+        if (request.socket.type === UtpSocketType.READ && request.socket.reader !== undefined) {
+            if (this.currentPacket.header.seqNr < request.socket.reader!.nextDataNr) {
+                this.logger.extend('PROCESS_CURRENT_PACKET')(`Packet ${this.currentPacket.header.seqNr} already processed.`)
+                this.currentPacket = this.packetHeap.pop()
+                return this.processCurrentPacket()            
+            } else if (this.currentPacket.header.seqNr > request.socket.reader!.nextDataNr) {
```

Comment:
This is what makes selective acks basically never happen, right?

Reply:
Right -- this still lacks a way to respond to actual packet loss.  This will just happen infinitely (or till other side sends a RESET).

Reply:
Commit 774053c adds some logic to process out of order packets if there are 3 or more packets for that request in the heap.  This preserves the reordering for the packet pair phenomenon, but still allows the `selective ack` response to happen if we get more than 3 packets ahead.